lib/app/dependency.go

```diff
@@ -17,106 +17,136 @@ limitations under the License.
-	state := &state{
-		visitedPackages: map[string]struct{}{},
-		visitedApps:     map[string]struct{}{},
-	}
-	if err = getDependencies(app, apps, state); err != nil {
+func GetDependencies(req GetDependenciesRequest) (result *Dependencies, err error) {
+	if err := req.checkAndSetDefaults(); err != nil {
 		return nil, trace.Wrap(err)
 	}
-	result = &Dependencies{}
-	result.Packages = append(result.Packages, state.packages...)
-	if state.runtimePackage != nil {
-		result.Packages = append(result.Packages, *state.runtimePackage)
+	state := &state{
+		packages: make(map[loc.Locator]struct{}),
+		apps:     make(map[loc.Locator]struct{}),
 	}
-	for _, locator := range state.apps {
-		if !locator.IsEqualTo(app.Package) {
-			result.Apps = append(result.Apps, locator)
-		}
+	if err = req.getDependencies(req.App, state); err != nil {
+		return nil, trace.Wrap(err)
 	}
-	result.Packages = loc.Deduplicate(result.Packages)
-	return result, nil
+	return &state.deps, nil
 }
 
-// VerifyDependencies verifies that the specified application app has all the dependent
-// packages available in the provided package service
-func VerifyDependencies(app *Application, apps Applications, packages pack.PackageService) error {
-	dependencies, err := GetDependencies(app, apps)
-	if err != nil {
-		return trace.Wrap(err)
-	}
-	for _, dependency := range dependencies.Packages {
-		_, err := packages.ReadPackageEnvelope(dependency)
-		if err != nil {
-			if trace.IsNotFound(err) {
-				log.Debugf("Package dependency %v is not present.", dependency)
-			}
-			return trace.Wrap(err)
-		}
-		log.Debugf("Package dependency %v is present.", dependency)
+// GetDependenciesRequest describes a request to transitively enumerate packages dependencies
+// for a specific application package
+type GetDependenciesRequest struct {
+	// App specifies the application to fetch dependencies for
+	App Application
+	// Apps specifies the application service
+	Apps Applications
+	// Pack specifies the package service
+	Pack pack.PackageService
+	// FieldLogger specifies the logger
+	log.FieldLogger
+}
+
+// AsPackages returns dependencies as a list of package identifiers
+func (r Dependencies) AsPackages() (result []loc.Locator) {
+	result = make([]loc.Locator, 0, len(r.Packages)+len(r.Apps))
+	for _, pkg := range r.Packages {
+		result = append(result, pkg.Locator)
 	}
-	for _, dependency := range dependencies.Apps {
-		_, err := apps.GetApp(dependency)
-		if err != nil {
-			if trace.IsNotFound(err) {
-				log.Debugf("App dependency %v is not present.", dependency)
-			}
-			return trace.Wrap(err)
-		}
-		log.Debugf("App dependency %v is present.", dependency)
+	for _, app := range r.Apps {
+		result = append(result, app.Package)
 	}
-	return nil
+	return result
+}
+
+// UniqPackages returns packages without duplicates
+func UniqPackages(packages []pack.PackageEnvelope) []pack.PackageEnvelope {
+	sort.Sort(packagesByLocator(packages))
```

Comment:
My uses are such that mutating the slices in-place is fine but I'll leave a note in the comment.
And if a future user is unhappy with such behavior, the API can receive a copy instead (otherwise it would be making a decision for the caller it might not like).